src/test/java/com/amazon/ion/bytecode/util/BytecodeBufferTest.kt

```diff
@@ -0,0 +1,712 @@
+
+        // Should be able to add new items starting from truncated size
+        buffer.add(99)
+        Assertions.assertEquals(99, buffer.get(2))
+        Assertions.assertEquals(3, buffer.size())
+    }
+
+    @Test
+    fun `truncate to zero makes buffer empty`() {
+        val buffer = BytecodeBuffer()
+        buffer.add(10)
+        buffer.add(20)
+
+        buffer.truncate(0)
+        Assertions.assertEquals(0, buffer.size())
+        Assertions.assertTrue(buffer.isEmpty())
+    }
+
+    @Test
+    fun `truncate with length greater than size throws IllegalArgumentException`() {
+        val buffer = BytecodeBuffer()
+        buffer.add(10)
+        buffer.add(20)
+
+        val exception = assertThrows<IllegalArgumentException> {
+            buffer.truncate(5)
+        }
+        Assertions.assertTrue(exception.message!!.contains("length exceeds number of values"))
+
+        // Buffer should remain unchanged after exception
+        Assertions.assertEquals(2, buffer.size())
+        Assertions.assertEquals(10, buffer.get(0))
+        Assertions.assertEquals(20, buffer.get(1))
+    }
+
+    @Test
+    fun `truncate allows truncating to current size`() {
+        val buffer = BytecodeBuffer()
+        buffer.add(10)
+        buffer.add(20)
+
+        // Should not change anything
+        buffer.truncate(2)
+        Assertions.assertEquals(2, buffer.size())
+        Assertions.assertEquals(10, buffer.get(0))
+        Assertions.assertEquals(20, buffer.get(1))
+    }
+
+    @Test
+    fun `buffer grows automatically when capacity is exceeded`() {
+        val buffer = BytecodeBuffer()
+        val initialCapacity = buffer.capacity()
+
+        // Add items beyond initial capacity
+        for (i in 0 until initialCapacity + 5) {
+            buffer.add(i)
+        }
+
+        Assertions.assertEquals(initialCapacity + 5, buffer.size())
+        Assertions.assertTrue(buffer.capacity() > initialCapacity)
+
+        // Verify all items are accessible
+        for (i in 0 until initialCapacity + 5) {
+            Assertions.assertEquals(i, buffer.get(i))
+        }
+    }
+
+    @Test
+    fun `toArray returns defensive copy with correct size`() {
+        val buffer = BytecodeBuffer()
+        buffer.add(10)
+        buffer.add(20)
+        buffer.add(30)
+
+        val array = buffer.toArray()
+        Assertions.assertEquals(3, array.size)
+        Assertions.assertEquals(10, array[0])
+        Assertions.assertEquals(20, array[1])
+        Assertions.assertEquals(30, array[2])
+
+        // Verify it's a defensive copy by modifying the returned array
+        array[0] = 999
+        Assertions.assertEquals(10, buffer.get(0)) // Original should be unchanged
+    }
+
+    @Test
+    fun `toArray returns empty array for empty buffer`() {
+        val buffer = BytecodeBuffer()
+        val array = buffer.toArray()
+        Assertions.assertEquals(0, array.size)
+    }
+
+    @Test
+    fun `unsafeGetArray returns backing array`() {
```

Comment:
This doesn't seem to actually test that it is the backing array, if that's important. You could modify a value at an index and then check...